rfcs/reorganize_src_lib.md

```diff
@@ -0,0 +1,117 @@
+
+In order to minimize merge conflict, moving the libraries to various subfolders can be done iteratively. And we could start by the things we are most confident about.
+
+At the moment, I would start with:
+
+- The `utils` folder: these libraries are quite easy to identify and moving them to a `utils` folder would unclutter the `src/lib` directory quite a bit.
+
+- The `crypto` folder (at least the libraries which were part of [this PR](https://github.com/MinaProtocol/mina/pull/9540)), and I see that this was already started in the
+`feature/group-lookups-mina` branch.
+
+## First attempt at grouping
+
+
+I tried to create meaningful groups and to not have to many edges in the following graph of [dependencies between groups](./res/reorganize_src_lib/class_graph.png) (which is still a bit complex).
+
+- The section below contains some information about the groups.
+- In the following dependency graphs, the libraries are colored according to their respective groups. The first one is the most readable (but contains less information) because it does not contain the `Utils` and `Logging` groups, and was run through the `tred` tool, which removes a lot of edges while preserving reachability.
+
+    - [without utils and logging - treded](./res/reorganize_src_lib/no_utils_tred.png)
+    - [without utils and logging](./res/reorganize_src_lib/no_utils.png)
+    - [all libraries - treded](./res/reorganize_src_lib/all_libs_tred.png)
+    - [all libraries](./res/reorganize_src_lib/all_libs.png)
+
+- The exact groups can also be found in the [library_group.md](./res/reorganize_src_lib/library_groups.md) file.
+
+
+
+### Current Groups
+
+- `Others`: Libraries that could stay at the toplevel, as their own group, or that I am not sure about.
+For instance the `mina_base` library that contains a lot a files and will maybe be split in the future.
+
+- `Utils`: Non application specific and usually small libraries. Most of the time leaves of the dependency graph.
+
+- `Logging`: Libraries related to logging that everyone can depend on (even the `Utils` one).
+
+- `Testing`: This folder would would contain two types of testing libraries:
+    - High level calls to tests such as `command_line_tests`.
+    - Test related utils such as `quickcheck_lib` (The difference with the `Utils` folder is that these would be allowed to depend on other application specific libraries).
+
+  The other folders could also contain a `testing` subfolder for specific test related libraries (for instance the `mina_net2_tests` would be in the `Network` group).
+    
+
+- `Processes`: Libraries related to parallelism and inter process communication,
+   such as `child_processes` and `pipe_lib`.
+
+- `Network`: Libraries related to network communication between the various component.
+
+- `Config`: Libraries such as `mina_numbers` defining primitive types used by most other groups.
```

Comment:
Libraries such as `genesis_constants` and `coda_runtime_config` were also in the `Config` group (alongside the ones that would now go in the new `Types` group) :+1: .

One downside of putting `coda_genesis_ledger`, `coda_genesis_proof` and `precomputed_values` in the same `Config` library is that they are higher in the dependency graph. So it would add edges to the [graph of groups](https://github.com/MinaProtocol/mina/blob/bd62f8816c3adfa7b00aa061ddd667c68f07968f/rfcs/res/reorganize_src_lib/class_graph.png). This is not always a bad thing but we would lose some precision with this high level view.

- `coda_genesis_ledger`: This one depends on mina_base, while the others do not.
Appart from that it is not too far so maybe after splitting `mina_base` this won't be a problem anymore ?

- `coda_genesis_proof`  and  `precomputed_values`: They depend on `mina_base` but also on `blockchain_snark` which is even higher up in the dependency graph. Because of this dependency, it is for now in the same group as `blockchain_snark`.
   
One way to have the best of both world, would be to effectively have subgroups in the `Config` group, which could go both way in terms of directory structure:

- Have multiple libraries in the `config` folder, so they are physically together but we can still separate them in the graph of groups if we want.
- Have `config` subfolders in some of the other groups (similarly to the optional `testing` one). So `coda_genesis_proof` and `precomputed_values` would go in a `config` subfolder of the group that contains `blockchain_snark` ?

In the second option the various dependency graphs would be more similar to the actual directory structure.

Reply:
> * `coda_genesis_proof`  and  `precomputed_values`: They depend on `mina_base` but also on `blockchain_snark` which is even higher up in the dependency graph. Because of this dependency, it is for now in the same group as `blockchain_snark`.

This is a good motivation for reworking these so that they no-longer depend on `blockchain_snark`. They haven't needed to for a while, it should just be a matter of doing that work.